### PR TITLE
Support promoting existing query as template query in dataspace

### PR DESCRIPTION
**Pull request by YannanGao-gs**
## Summary

Support promoting existing query as template query in dataspace

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

https://github.com/finos/legend-studio/assets/73408381/a3696f70-6dc5-4e5a-8f13-54dcb6be6fe1

### CODE REVIEW

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+        (yield this.graphManagerState.graphManager.getQueryInfo(
+          query.id,
+        )) as QueryInfo;
+      this.currentQueryProject = StoreProjectData.serialization.fromJson(
+        (yield this.depotServerClient.getProject(
+          this.currentQuery.groupId,
+          this.currentQuery.artifactId,
+        )) as PlainObject<StoreProjectData>,
+      );
+      this.setWorkspaceName(`${DEFAULT_WORKSPACE_NAME_PREFIX}-${query.id}`);
```
Comment (MauricioUyaguari) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:327]: lets test this workspace name, sometimes gitlab build fails with workspace names like these

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts [hunk 2]
```diff
@@ -0,0 +1,666 @@
...
+            this.currentQuery.artifactId,
+            this.currentQuery.versionId,
+          ),
+          this.sdlcServerClient.getConfiguration(
+            this.currentQueryProject.projectId,
+            undefined,
+          ),
+        ])) as [Entity[], PlainObject<ProjectConfiguration>];
+
+        const [currentProjectEntities, currentProjectConfiguration] = [
```
Comment (MauricioUyaguari) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:265]: I don't think we should build the full graph here. Even the system graph, unless we have a reason?
We should maybe create a light Dataspace and jsut buidl that from the Entity. 
So fetch the entity directly. 
We will do the compiler checks once they click procceed.

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+    try {
+      this.currentProjectConfigurationStatus =
+        (yield fetchProjectConfigurationStatus(
+          project.projectId,
+          undefined,
+          this.applicationStore,
+          this.sdlcServerClient,
+        )) as ProjectConfigurationStatus;
+
+      const workspacesInConflictResolutionIds = (
```
Comment (MauricioUyaguari) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:364]: lets save on these api calls unless necessary, do we need these?

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+      this.applicationStore.alertService.setBlockingAlert({
+        message: `Fetching and updating project...`,
+        prompt: 'Please do not close the application',
+        showLoading: true,
+      });
+      const dataSpaceExecutableTemplate = new DataSpaceExecutableTemplate();
+      dataSpaceExecutableTemplate.id = this.templateQueryId;
+      dataSpaceExecutableTemplate.title = this.templateQueryTitle;
+      dataSpaceExecutableTemplate.description = this.templateQueryDescription;
+      dataSpaceExecutableTemplate.executionContextKey = guaranteeNonNullable(
```
Comment (MauricioUyaguari) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:426]: lets add a TODO here, we need to update this code once we properly store existing dataspace quieres, they should not have mapping/runtime but executionKeys.

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+            this.editorStore.projectConfigurationEditorState
+              .containsSnapshotDependencies;
+          const isCreateReviewDisabled =
+            Boolean(workspaceReviewState.workspaceReview) ||
+            workspaceContainsSnapshotDependencies ||
+            !workspaceReviewState.canCreateReview ||
+            workspaceReviewState.sdlcState.isActiveProjectSandbox;
+          workspaceReviewState.reviewTitle =
+            'code review - promote query as a template query to dataspace';
+          if (!isCreateReviewDisabled) {
```
Comment (MauricioUyaguari) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:543]: dont think we notify users if createReviewIsDiabled.

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts [hunk 6]
```diff
@@ -0,0 +1,666 @@
...
+        : [dataSpaceExecutableTemplate];
+
+      // 2. check if the graph compiles properly
+      this.applicationStore.alertService.setBlockingAlert({
+        message: `Checking workspace compilation status...`,
+        prompt: 'Please do not close the application',
+        showLoading: true,
+      });
+
+      const graphGrammar =
```
Comment (MauricioUyaguari) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:451]: lets avoid the studio graph here. we can add methods to pure graph manager if need be. 
but we should be able to do patch the dataspace and add the template query without building the graph. 
We can still compile with engine with the full patched entities but we shouldnt need to build our graph for this.

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts (hunk 2 above)
Comment (YannanGao-gs) [reply packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceTemplateQueryPromotionReviewerStore.ts:265]: I did try to build light graph at first, however, light graph won't generate existing dataspace's executables  e.g. 
Screenshot 2024-04-15 at 9 39 42 AM

Then it has to append newly added executable entity plain object to original dataspace entity

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**